overlay.d/05rhcos/usr/lib/dracut/modules.d/40rhcos-fips/rhcos-fips.sh

```diff
@@ -59,9 +59,15 @@ firstboot() {
     mkdir -p "${tmpsysroot}/boot"
     mount /dev/disk/by-label/boot "${tmpsysroot}/boot"
 
+    # https://github.com/dracutdevs/dracut/blob/master/modules.d/01fips/fips.sh#L146
+    # Dracut looks for .vmlinux-xyz.hmac, which in our case is installed not under /boot,
+    # but under /boot/ostree/rhcos-xxxyyyzz/. To avoid installation failure we have to
+    # specify BOOT_IMAGE karg.
```

Comment:
On GRUB-supported platforms, the `BOOT_IMAGE` arg is already injected by GRUB into the command-line so this should Just Work. For s390x, we should check how the dracut code is intended to work, since this is something that traditional RHEL would hit as well.

Reply:
Here comes interesting part - `zipl` by default sets `BOOT_IMAGE=0` (https://github.com/ibm-s390-linux/s390-tools/blob/master/zipl/boot/menu.c#L188). And because we don't use `grub`, `dracut` fails. 
Also there is some strange issue during `cosa build` : 
```
dracut: *** Including module: fips ***
mknod: /tmp/dracut/dracut.yRQ17k/initramfs/dev/random: Operation not permitted
dracut: Cannot create /dev/random
dracut: To create an initramfs with fips support, dracut has to run as root
```

Reply:
Also i wonder why https://github.com/ostreedev/ostree/pull/1962/ doesn't work.

```
[    5.802578] dracut-pre-trigger[719]: Mounting /dev/disk/by-label/boot as /boot 
[    5.841955] SGI XFS with ACLs, security attributes, no debug enabled 
[    5.876049] XFS (dasda2): Mounting V5 Filesystem 
[    5.882649] EXT4-fs (dasda1): mounted filesystem with ordered data mode. Opts: (null) 
[    5.883933] dracut-pre-trigger[719]: Checking integrity of kernel 
[    5.886162] dracut: FATAL: FIPS integrity test failed 
[    5.886166] dracut: Refusing to continue 
[    5.886269] dracut-pre-trigger[719]: Warning: /boot//.vmlinuz-4.18.0-240.22.1.el8_3.s390x.hmac does not exist 
[    5.886280] dracut-pre-trigger[719]: Warning: /boot//.vmlinuz-4.18.0-240.22.1.el8_3.s390x.hmac does not exist 
[    5.917621] XFS (dasda2): Ending clean mount 
[    5.917951] XFS (dasda2): Quotacheck needed: Please wait. 
[    5.982502] systemd[1]: Shutting down. 
```


Please correct me if i'm wrong - `rpm-ostree` doesn't create anything during install (because we just copy `osmet` image) and later during `firstboot` . Or should it create those `*.hmac` during `cosa build***` ?

Reply:
>  Also there is some strange issue during cosa build :

We can ignore those issues, see: https://github.com/coreos/rpm-ostree/pull/1951

> Here comes interesting part - zipl by default sets BOOT_IMAGE=0 (https://github.com/ibm-s390-linux/s390-tools/blob/master/zipl/boot/menu.c#L188). And because we don't use grub, dracut fails.
> Also i wonder why ostreedev/ostree#1962 doesn't work.

Ahh OK, I think I see what's going on here.

So, on OSTree systems, the kernel, HMAC file, and initramfs are placed under `/boot/ostree/$name-$hash/`.

On GRUB systems, the `BOOT_IMAGE` is a path to the kernel image, and the dracut FIPS code uses that to also find the HMAC file.

It looks like on s390x, the `BOOT_IMAGE` is an integer representing the index of the configuration that was booted into? Does this roughly correspond to the `version` field of the BLS config?

Anyway, I think what's going on here is that dracut doesn't know how to use this field and so just ignores it and falls back to trying it in the root of `/boot`. This works fine on traditional systems because the kernel and HMAC files are indeed placed at the root of the boot partition. So this only breaks on OSTree systems.

Hmm, ideally we would teach dracut how to use this index to figure out the kernel path from it (and so the HMAC path). Do you think mapping the index back to the BLS is feasible or is it too error-prone? I suspect there might not actually be any traces of where the configuration came from once it's written out by `zipl` (or maybe there's some extra metadata we can shove in there?). /cc @martinezjavier

I don't think we should pre-inject `BOOT_IMAGE` like this patch is doing. Definitely wouldn't want to do this on GRUB platforms, since it contains dynamic information about e.g. which devices was actually used. But even on s390x, it's possible that there are tools which expect an integer there for their purposes.

Reply:
>So, on OSTree systems, the kernel, HMAC file, and initramfs are placed under /boot/ostree/$name-$hash/.

exactly. and becuase 
>on s390x, the BOOT_IMAGE is an integer representing the index of the configuration that was booted

i had to modify BOOT_IMAGE + https://github.com/nikita-dubrovskii/s390-tools/commit/70dde0e7d64f39a6f4747306c15ebc053ba01b19 . I've asked `zipl` experts about this, maybe it would be possible to mimic grub here.

So with all those patches system runs (rescue shell) with `fips`, but `ignition` is skipped. 
Could `fips` affect `ignition` ? I use my own simple `rhcos.ign` augmented with https://github.com/coreos/coreos-assembler/blob/master/mantle/kola/tests/fips/fips.go#L25

Reply:
> I've asked zipl experts about this, maybe it would be possible to mimic grub here.

Oh that would be the ideal fix if possible!

> So with all those patches system runs (rescue shell) with `fips`, but `ignition` is skipped.

Hmm, not sure offhand. Can you paste all the logs in e.g. a gist? You should see the system boot, run `ignition-fetch`, inject `fips=1 boot=LABEL=boot`, and then immediately reboot (and then run through all the Ignition stages).

Reply:
Discussion about `zipl` is going on. But looks like we can not drop `BOOT_IMAGE=N` because some further tooling is relying on this. I've found why `ignition` was skipped and made an update, so now system runs thru all ignition steps, but anyhow i end up with a rescue shell. Here is an installation log - https://bugzilla.redhat.com/show_bug.cgi?id=1955621#c8 , maybe i've missed smth?

Reply:
Here is a link to zipl BOOT_IMAGE issue - https://github.com/ibm-s390-linux/s390-tools/issues/78

Reply:
There's an even more obvious practical reason why hardcoding `BOOT_IMAGE` won't work: it'll become out of date as soon as the node upgrades to a new image with a different kernel.

@Prashanth684 Do you know the answer to

> Hmm, ideally we would teach dracut how to use this index to figure out the kernel path from it (and so the HMAC path). Do you think mapping the index back to the BLS is feasible or is it too error-prone?

Reply:
right now i see that BOOT_IMAGE is set to 0. i tried an rhcos upgrade and didn't see it change so not sure if that will help us. @nikita-dubrovskii does BOOT_IMAGE update to 1 when you do an upgrade?

Reply:
From my naive reading of https://github.com/ibm-s390-linux/s390-tools/blob/master/zipl/boot/menu.c, I think it represents the index of the menu entry that was selected, so 0 is expected. The question is whether we can go from there and map it back to e.g. the vmlinuz path. Because if so, then we can teach dracut to use it.

Reply:
oh..right...yeah it's always 0.  not sure if there is a good way to map it to the kernel/initramfs.let me ask around.

Reply:
so after talking to a few folks looks like there is no good way to map the index back to the BLS entry.

Reply:
In that case, I think I agree that https://github.com/ibm-s390-linux/s390-tools/issues/78 is the best way to solve this, even if it happens at "boot record writing" time. If we can't change the existing semantics of `BOOT_IMAGE=N`, then maybe best to have a separate variable for this, e.g. `BOOT_IMAGE_PATH`.

One hack we could do is to have ostree on s390x symlink all the kernel and HMAC files from the root of the boot partition, but... yuck. (And you get into the question of what to do if somehow two `vmlinuz` files named the same across deployments somehow don't have the same checksum.)

Reply:
Let's chat in https://github.com/ibm-s390-linux/s390-tools/issues/78.